util/compression/compression.go

```diff
@@ -241,3 +234,72 @@ func ConvertAllLayerMediaTypes(oci bool, descs ...ocispecs.Descriptor) []ocispec
 	}
 	return converted
 }
+
+func decompress(ctx context.Context, cs content.Store, desc ocispecs.Descriptor) (r io.ReadCloser, err error) {
+	ra, err := cs.ReaderAt(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	esgz, err := EStargz.Is(ctx, cs, desc.Digest)
+	if err != nil {
+		return nil, err
+	} else if esgz {
+		r, err = decompressEStargz(io.NewSectionReader(ra, 0, ra.Size()))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		r, err = cdcompression.DecompressStream(io.NewSectionReader(ra, 0, ra.Size()))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &readCloser{r, ra.Close}, nil
+}
+
+type readCloser struct {
+	io.ReadCloser
+	closeFunc func() error
+}
+
+func (rc *readCloser) Close() error {
+	err1 := rc.ReadCloser.Close()
+	err2 := rc.closeFunc()
+	if err1 != nil {
+		return errors.Wrapf(err1, "failed to close: %v", err2)
+	}
+	return err2
+}
+
+type WriteCloser struct {
+	io.WriteCloser
+	CloseFunc func() error
+}
+
+func (wc *WriteCloser) Close() error {
+	err1 := wc.WriteCloser.Close()
+	err2 := wc.CloseFunc()
+	if err1 != nil {
+		return errors.Wrapf(err1, "failed to close: %v", err2)
+	}
+	return err2
+}
```

Comment:
Looks like unrelated to compression. We can introduce another pkg maybe.

Reply:
Fixed by introducing a new package `github.com/moby/buildkit/util/iohelper`.

We can move the codes like `readCloser`, `nopWriteCloser` in the whole project into this package (I will do this but not in this PR).